File: Data Structures & Algorithms/trapping-rain-water/submission-4.java
class Solution {
    public int trap(int[] height) {
    int left = 0;
    int right = height.length-1; 

    int leftMax = height[0]; 
    int rightMax = height[height.length-1]; 
    int result = 0; 

    while(left < right){ 
        if(leftMax <= rightMax){ 
            left++; 
            result += Math.max(0,leftMax - height[left]);
            leftMax = Math.max(leftMax, height[left]); 
            
        } else { 
            right--; 
            result +=  Math.max(0, rightMax - height[right]);
            rightMax = Math.max(rightMax, height[right]); 
             // no need to take minimum of either side.
        }
    }


    return result;
    }
}

File: Data Structures & Algorithms/binary-search/submission-4.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0 
        right = len(nums) -1 

        while(left <= right): 
            mid = left + (right - left)/2 

            if(nums[mid] < target): 
                left = mid + 1 
                continue
            elif(nums[mid] > target): 
                right = mid -1 
                continue
            else: 
                return mid
        
        return -1

            